cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/AbstractProgramRuntimeService.java

```diff
@@ -71,11 +90,91 @@ public RuntimeInfo run(Program program, ProgramOptions options) {
     ProgramRunner runner = programRunnerFactory.create(ProgramRunnerFactory.Type.valueOf(program.getType().name()));
     Preconditions.checkNotNull(runner, "Fail to get ProgramRunner for type " + program.getType());
     ProgramOptions optionsWithRunId = addRunId(options, RunIds.generate());
-    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithRunId), program);
-    programStarted(runtimeInfo);
+    File tempDir = createTempDirectory(program.getId());
+    Runnable cleanUpTask = createCleanupTask(tempDir);
+    ProgramOptions optionsWithPlugins = createPluginSnapshot(optionsWithRunId, program.getId().getNamespace(), tempDir,
+                                                             program.getApplicationSpecification());
+    final RuntimeInfo runtimeInfo = createRuntimeInfo(runner.run(program, optionsWithPlugins), program);
+    programStarted(runtimeInfo, cleanUpTask);
     return runtimeInfo;
   }
 
+  private Runnable createCleanupTask(final File... resources) {
+    return new Runnable() {
+      @Override
+      public void run() {
+        for (File file : resources) {
+          if (file == null) {
+            continue;
+          }
+
+          try {
+            if (file.isDirectory()) {
+              DirUtils.deleteDirectoryContents(file);
+            } else {
+              file.delete();
+            }
+          } catch (Throwable t) {
+            LOG.warn("Exception when cleaning up resource {}", file, t);
+          }
+        }
+      }
+    };
+  }
+
+  /**
+   * Creates a local temporary directory for this program run.
+   */
+  private File createTempDirectory(Id.Program programId) {
+    File tempDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+                            cConf.get(Constants.AppFabric.TEMP_DIR)).getAbsoluteFile();
+    File dir = new File(tempDir, String.format("%s.%s.%s.%s.%s",
+                                               programId.getType().name().toLowerCase(),
+                                               programId.getNamespaceId(), programId.getApplicationId(),
+                                               programId.getId(), UUID.randomUUID().toString()));
```

Comment:
BTW, the MapReduceRuntimeService is using the same directory (if using runid). I think it's better for MapReduceRuntimeService to create a sub-dir under the tempDir. Something like `data/temp/runner/mapreduce.appId....`